projects/Grandine-Implementing-Tokio-Tracing-For-Debugging-And-Performance-Analysis.md

```diff
@@ -0,0 +1,303 @@
+    style E fill:#d5f0d5
+    style G fill:#f8d7da
+
+```
+
+Tokio Console: A real-time debugging and diagnostics tool for Tokio applications. It gives you a live view of all your tasks, resources, and their behavior. To use it, you need to add the console-subscriber and enable specific tokio trace features.
+
+Flamegraphs with tracing-flame: For performance profiling, you can use tracing-flame and flamegraph to generate flamegraphs. These are visualizations that show exactly where your program is spending its time, which is invaluable for identifying performance bottlenecks.
+
+By integrating Tokio Tracing, you move from simple text-based logging to a powerful, structured diagnostics system that provides deep insights into both the correctness and performance of your application.
+
+---
+
+## Specification
+
+The implementation will deliver the following technical outcomes:
+
+- Tracing Integration
+  - `tracing_subscriber` set as the main diagnostics backend.
+  - Unified configuration for console logs, JSON output, and optional flamegraph export.
+  - Runtime filtering via environment variables.
+- Instrumentation
+  - HTTP API handlers instrumented to log request lifecycle and parameters.
+  - P2P networking spans for peer events and message processing with structured metadata.
+  - Fork choice and validator duties instrumented to trace consensus-critical workflows.
+- Advanced Tooling
+  - Integration with `tokio-console` for live monitoring of tasks.
+  - Optional `tracing-flame` support for performance profiling.
+- Configurability
+  - CLI or config-based toggles for enabling or disabling tracing features.
+  - Distinction between production and development modes.
+- Documentation
+  - Guides and examples for configuring tracing and interpreting outputs.
+
+
+---
+
+## Roadmap
+
+The project will be implemented in four incremental phases, progressing from foundational setup to advanced tooling and full integration across the Grandine codebase.
+
+
+```mermaid
+flowchart TD
+    style A fill:#e6f7ff,stroke:#1890ff,stroke-width:2px
+    style B fill:#e6f7ff,stroke:#1890ff,stroke-width:2px
+    style C fill:#e6f7ff,stroke:#1890ff,stroke-width:2px
+    style D fill:#e6f7ff,stroke:#1890ff,stroke-width:2px
+
+    A[Phase 1: <br>Foundational Setup]
+    B[Phase 2: <br>Network Instrumentation]
+    C[Phase 3: Core Logic <br>Instrumentation]
+    D[Phase 4: <br> Advanced Tooling <br>& FInal Migration]
+
+    A --> B --> C --> D
+
+    subgraph S1 [Phase 1 Details]
+        A1(Enable tracing)
+        A2(Replace logger)
+    end
+    A --> S1
+
+    subgraph S2 [Phase 2 Details]
+        B1(Trace HTTP API)
+        B2(Trace P2P)
+    end
+    B --> S2
+
+    subgraph S3 [Phase 3 Details]
+        C1(Fork Choice)
+        C2(Validator Duties)
+    end
+    C --> S3
+
+    subgraph S4 [Phase 4 Details]
+        D1(Tokio Console)
+        D2(Flamegraphs)
+        D3(Final migration)
+    end
+    D --> S4
+
+```
+
+
+### Phase 1: Foundational Setup
```

Comment:
it feels redundant to have both chart and description of each phase. Choose just one please